File: C_plus_plus/Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *head = NULL,*p;
        int carry = 0;
        while(l1&&l2){
            if(!head){
                head = new ListNode((l1->val+l2->val+carry)%10);
                carry = (l1->val+l2->val+carry)/10;
                head->next=NULL;
                p = head;
            }
            else{
                p->next = new ListNode((l1->val+l2->val+carry)%10);
                carry = (l1->val+l2->val+carry)/10;
                p = p->next;
                p->next = NULL;
            }
            l1 = l1->next;
            l2 = l2->next;
        }
        while(l1){
            p->next = new ListNode((l1->val+carry)%10);
            carry = (l1->val+carry)/10;
            p = p->next;
            p->next = NULL; 
            l1 = l1->next;
        }
        while(l2){
            p->next = new ListNode((l2->val+carry)%10);
            carry = (l2->val+carry)/10;
            p = p->next;
            p->next = NULL; 
            l2 = l2->next;
        }
        if(carry){
            p->next = new ListNode(carry);
            p->next->next = NULL;
        }
        return head;
    }
};

File: java/Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = null, p = null;
        int carry = 0;
        while (l1 != null || l2 != null || carry > 0){
            if (result == null) {
                result = new ListNode(0);
                p = result;
            }
            else {
                p.next = new ListNode(0);
                p = p.next;
            }
            int sum = carry;
            if(l1 != null) {
                sum += l1.val;
                l1 = l1.next ;
            }
            if(l2 != null){
                sum += l2.val;
                l2 = l2.next ;
            }
            p.val = sum % 10;
            carry = sum / 10;
        }
        return result;
    }
}
